### PR TITLE
Add OED estimators and accompanying tests

### PR DESCRIPTION
**Pull request by ae-foster**
This PR adds the bare bones of the new estimators for EIG and the tests for those methods. It does not change any examples (beyond deleting a couple out-dated examples). I know it weighs in at 1.5k of changes - so sorry, I did my best to keep this as small as possible without insane rewriting of the code.

Summary of changes at a code level:
 - `pyro/contrib/oed/eig.py` includes the bulk of the changes for estimators
 - `tests/contrib/oed/` includes the new tests
 - glmm is mostly compatibility changes, plus @eb8680 's Laplace guide
 - everything else (util files, etc.) is mostly bookkeeping

### CODE REVIEW

File: pyro/contrib/oed/eig.py [hunk 1]
```diff
@@ -327,14 +463,180 @@ def loss_fn(design, num_particles):
...
+        if evaluation:
+            trace.compute_log_prob()
+            terms += sum(trace.nodes[l]["log_prob"] for l in observation_labels)
+
+        return safe_mean_terms(terms)
+
+    return loss_fn
+
+
+def safe_mean_terms(terms):
```
Comment (eb8680) [inline pyro/contrib/oed/eig.py]: nit: if this isn't used elsewhere and isn't meant to be called by users, rename to `_safe_mean_terms`

File: pyro/contrib/glmm/glmm.py [hunk 2]
```diff
@@ -237,11 +247,11 @@ def bayesian_linear_model(design, w_means={}, w_sqrtlambdas={}, re_group_sizes={
         elif response == "bernoulli":
             return pyro.sample(response_label, dist.Bernoulli(logits=prediction_mean).to_event(1))
         elif response == "sigmoid":
-            base_dist = dist.Normal(prediction_mean, obs_sd).to_event(1)
             # You can add loc via the linear model itself
             k = k.expand(prediction_mean.shape)
-            transforms = [AffineTransform(loc=torch.tensor(0.), scale=k), SigmoidTransform()]
-            response_dist = dist.TransformedDistribution(base_dist, transforms)
+            response_dist = dist.CensoredSigmoidNormal(
```
[the remark this answers is not in the document]
Comment (ae-foster) [reply pyro/contrib/glmm/glmm.py]: Indeed- I'll revert to the old method and add the new method in a later PR

File: tests/contrib/oed/test_linear_models_eig.py [hunk 3]
```diff
@@ -0,0 +1,197 @@
...
+import pyro.distributions as dist
+import pyro.optim as optim
+from pyro.infer import Trace_ELBO
+from pyro.contrib.glmm import known_covariance_linear_model
+from pyro.contrib.oed.util import linear_model_ground_truth
+from pyro.contrib.oed.eig import (
+    nmc_eig, posterior_ape, marginal_eig, marginal_likelihood_eig, vnmc_eig, laplace_vi_ape, lfire_eig,
+    donsker_varadhan_eig)
+from pyro.contrib.util import rmv, rvv
+from pyro.contrib.glmm.guides import LinearModelLaplaceGuide
```
[the remark this answers is not in the document]
Comment (ae-foster) [reply tests/contrib/oed/test_linear_models_eig.py]: So I wrote fresh guides for the tests to make them more self-contained & readable. These guides get to be way simpler because they do not have to be prepared to all kinds of models. I could also use the `glmm/guides` versions though if you think it would be better

**Comment by martinjankowiak** [comment]
@ae-foster thanks for the PR! haven't had a chance to look at this in detail yet but i wanted to flag a usability concern i ran into. basically all the estimators should return actual EIGs (by default, or at least optionally) and not just up to a design-independent constant. this at least wasn't the case for `posterior_ape = posterior_eig`. thoughts?

**Comment by ae-foster** [comment]
Right now, you have `posterior_ape`, `posterior_eig_lm` (use posterior_ape and then compute the prior entropy analytically using a Gaussian assumption) and `posterior_eig_mc` (same, but compute prior entropy using Monte Carlo). If you want, we could rename as `_posterior_ape` to indicate you shouldn't use that one?

**Comment by martinjankowiak** [comment]
hmm well wouldn't something roughly like this be cleanest?
```python
def posterior_eig(include_prior_entropy=True):
     if  include_prior_entropy:
        try:
             # analytic KL
        except:
            # MC
   _posterior_ape()
    .....```

**Comment by ae-foster** [comment]
Yeah I can refactor it like that

File: pyro/contrib/oed/eig.py [hunk 4]
```diff
@@ -327,14 +463,180 @@ def loss_fn(design, num_particles):
...
+            terms += sum(cond_trace.nodes[l]["log_prob"] for l in observation_labels)
+        else:
+            terms -= sum(cond_trace.nodes[l]["log_prob"] for l in observation_labels)
+
+        return safe_mean_terms(terms)
+
+    return loss_fn
+
+
+def lfire_loss(model_marginal, model_conditional, h, observation_labels, target_labels):
```
[the remark this answers is not in the document]
Comment (ae-foster) [reply pyro/contrib/oed/eig.py]: I'm not sure you would use this directly- you can always effectively use it directly by setting `num_steps=0` in `lfire_eig`

File: pyro/contrib/oed/eig.py [hunk 5]
```diff
@@ -327,14 +463,180 @@ def loss_fn(design, num_particles):
...
+            return safe_mean_terms(terms)
+
+        else:
+            h_joint = h(expanded_design, model_conditional_trace, observation_labels, target_labels)
+            return safe_mean_terms(h_joint)
+
+    return loss_fn
+
+
+def vnmc_eig_loss(model, guide, observation_labels, target_labels):
```
[the remark this answers is not in the document]
Comment (ae-foster) [reply pyro/contrib/oed/eig.py]: As with `lfire`. Maybe I should make these `_` functions?

**Comment by martinjankowiak** [comment]
if it helps i could write some sort of analytic KL helper along the lines of what's [here](https://github.com/pyro-ppl/pyro/blob/29bd73f476b8b9173eda3a8de71d5992888197f9/pyro/infer/trace_mean_field_elbo.py#L100)

**Comment by ae-foster** [comment]
We should be able to re-use the old `mean_field_guide_entropy` function that we wrote before (assuming that the prior is indeed separable)

File: tests/contrib/oed/test_linear_models_eig.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (eb8680) [reply tests/contrib/oed/test_linear_models_eig.py]: I think what you have here is fine, was just curious

File: pyro/contrib/glmm/guides.py [hunk 6]
```diff
@@ -6,35 +6,49 @@
...
+        self.regressor = {l: nn.Parameter(
+                regressor_init * torch.ones(*(d + (p, sum(y_sizes.values()))))) for l, p in w_sizes.items()}
         self.scale_tril = {l: nn.Parameter(
-                scale_tril_init*torch.ones(d, p, p)) for l, p in w_sizes.items()}
+                scale_tril_init * lexpand(torch.eye(p), *d)) for l, p in w_sizes.items()}
         # This registers the dict values in pytorch
         # Await new version to use nn.ParamterDict
-        self._registered = nn.ParameterList(self.scale_tril.values())
+        self._registered_re = nn.ParameterList(self.regressor.values())
+        self._registered_st = nn.ParameterList(self.scale_tril.values())
```
Comment (eb8680) [inline pyro/contrib/glmm/guides.py]: Do you want to switch to [`torch.nn.ParameterDict`](https://pytorch.org/docs/stable/nn.html#torch.nn.ParameterDict) here now that it's available in 1.1?

File: pyro/contrib/oed/eig.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (eb8680) [reply pyro/contrib/oed/eig.py]: If they aren't used outside this file, you can prefix them with `_`.  You should also add an `__all__` list to this file exporting only the things you expect to be called by users.

File: pyro/contrib/oed/eig.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (eb8680) [reply pyro/contrib/oed/eig.py]: Makes sense, add a one-sentence docstring to that effect?

File: pyro/contrib/glmm/glmm.py [hunk 7]
```diff
@@ -358,7 +368,22 @@ def analytic_posterior_cov(prior_cov, x, obs_sd):
...
+    Given a list of tensor sizes, returns the size of the resulting broadcasted
+    tensor.
+    Args:
+        shapes (list of torch.Size): list of tensor sizes
+    Copied from PyTorch 0.4.0
+    """
+    shape = torch.Size()
+    for s in shapes:
+        shape = torch._C._infer_size(s, shape)
+    return shape
```
[the remark this answers is not in the document]
Comment (neerajprad) [reply pyro/contrib/glmm/glmm.py]: There is a `torch.broadcast_tensors(*tensors)` which will directly broadcast all the tensors, so it will likely need some code modification. I remember @fritzo made a change to torch distributions to replace `torch._C._infer_size` with `torch.broadcast_tensors` for compatibility with JIT. Unless that is fixed, this will likely create issues with JIT.

[the remark this answers is not in the document]
Comment (ae-foster) [reply pyro/contrib/glmm/glmm.py]: I'll rewrite to use `torch.broadcast_tensors`- this code was probably for compatibility with a now-very-outdated version to torch

File: pyro/contrib/glmm/guides.py (hunk 6 above)
Comment (ae-foster) [reply pyro/contrib/glmm/guides.py]: Oh yes

File: pyro/contrib/oed/eig.py [hunk 8]
```diff
@@ -1,23 +1,105 @@
...
+    "nmc_eig",
+    "donsker_varadhan_eig",
+    "posterior_eig",
+    "marginal_eig",
+    "lfire_eig",
+    "vnmc_eig"
+]
+
+
+def laplace_vi_eig(model, design, observation_labels, target_labels, guide, loss, optim, num_steps,
```
Comment (martinjankowiak) [inline pyro/contrib/oed/eig.py]: why is this called `vi`?

File: pyro/contrib/oed/eig.py [hunk 9]
```diff
@@ -106,46 +217,61 @@ def naive_rainforth_eig(model, design, observation_labels, target_labels=None,
...
         observation_labels = [observation_labels]
     if isinstance(target_labels, str):
         target_labels = [target_labels]
 
-    # Take N samples of the model
-    expanded_design = lexpand(design, N)
-    trace = poutine.trace(model).get_trace(expanded_design)
-    trace.compute_log_prob()
+    s = 0.
+    if N_seq > 1:
```
Comment (martinjankowiak) [inline pyro/contrib/oed/eig.py]: what's `N_seq`? this and `independent_priors` missing from docstring

File: pyro/contrib/oed/eig.py [hunk 10]
```diff
@@ -46,24 +128,46 @@ def vi_ape(model, design, observation_labels, target_labels,
...
+    :param bool eig: Whether to compute the EIG or the average posterior entropy (APE). The EIG is given by
+                     `EIG = prior entropy - APE`. If `True`, the prior entropy will be estimated analytically,
+                     or by Monte Carlo as appropriate for the `model`. If `False` the APE is returned.
+    :param dict prior_entropy_kwargs: parameters for estimating the prior entropy.
+    :return: EIG estimate
     :rtype: `torch.Tensor`
 
     """
 
+    warnings.warn("`vi_eig` is deprecated in favour on the amortized version: `posterior_eig`.", DeprecationWarning)
```
Comment (martinjankowiak) [inline pyro/contrib/oed/eig.py]: typo: on => of

File: pyro/contrib/oed/eig.py [hunk 11]
```diff
@@ -231,14 +358,222 @@ def barber_agakov_ape(model, design, observation_labels, target_labels,
         `design`.
     :param int final_num_samples: The number of samples to use at the final evaluation, If `None,
         uses `num_samples`.
-    :return: EIG estimate, optionally includes full optimisatio history
+    :param bool eig: Whether to compute the EIG or the average posterior entropy (APE). The EIG is given by
+                 `EIG = prior entropy - APE`. If `True`, the prior entropy will be estimated analytically,
+                 or by Monte Carlo as appropriate for the `model`. If `False` the APE is returned.
+    :param dict prior_entropy_kwargs: parameters for estimating the prior entropy.
```
Comment (martinjankowiak) [inline pyro/contrib/oed/eig.py]: namely?

File: pyro/contrib/oed/eig.py [hunk 12]
```diff
@@ -231,14 +358,222 @@ def barber_agakov_ape(model, design, observation_labels, target_labels,
...
+    if return_history:
+        return out[0], out[1].sum(0) / num_theta_samples
+    else:
+        return out.sum(0) / num_theta_samples
+
+
+def vnmc_eig(model, design, observation_labels, target_labels,
+             num_samples, num_steps, guide, optim, return_history=False,
+             final_design=None, final_num_samples=None):
+    """Estimates the EIG using Variational Nested Monte Carlo (VNMC). The VNMC estimate is
```
Comment (martinjankowiak) [inline pyro/contrib/oed/eig.py]: here and elsewhere you could you some foster et al references?

File: pyro/contrib/oed/eig.py [hunk 13]
```diff
@@ -231,14 +358,222 @@ def barber_agakov_ape(model, design, observation_labels, target_labels,
...
+    .. math::
+
+        \\frac{1}{N}\\sum_{n=1}^N \\left[ \\log p(y_n | \\theta_n, d) -
+         \\log \\left(\\frac{1}{M}\\sum_{m=1}^M \\frac{p(\\theta_{mn})p(y_n | \\theta_{mn}, d)}
+         {q(\\theta_{mn} | y_n)} \\right) \\right]
+
+    where :math:`q(\\theta | y)` is the learned variational posterior approximation and
+    :math:`\\theta_n, y_n \\sim p(\\theta, y | d)`, :math:`\\theta_{mn} \\sim q(\\theta|y=y_n)`.
+
+    As :math:`N \\to \\infty` this is an upper bound on EIG. We minimise this upper bound by stochastic gradient.
```
Comment (martinjankowiak) [inline pyro/contrib/oed/eig.py]: by stochastic gradient descent?

File: pyro/contrib/oed/eig.py (hunk 8 above)
Comment (ae-foster) [reply pyro/contrib/oed/eig.py]: Let's change it to `laplace_eig`

File: pyro/contrib/oed/eig.py (hunk 9 above)
Comment (ae-foster) [reply pyro/contrib/oed/eig.py]: I'll document it... `N_seq > 1` actually doesn't work rn

File: pyro/contrib/oed/eig.py (hunk 12 above)
Comment (ae-foster) [reply pyro/contrib/oed/eig.py]: I will add a Foster et al. reference to the four estimators mentioned in our paper

File: pyro/contrib/oed/eig.py (hunk 9 above)
Comment (martinjankowiak) [reply pyro/contrib/oed/eig.py]: remove it?

Comment (ae-foster) [reply pyro/contrib/oed/eig.py]: yes that's sensible, we'll keep it on another branch for now

**Comment by ae-foster** [comment]
@martinjankowiak @eb8680 I think I've addressed all your comments now. I also removed `eig_estimation_benchmarking`: it was going to require a substantial rewrite, only to be changed again in the next PR. In later PRs we should add more user-friendly examples in `examples/contrib/oed` and maybe put numerical experiments for the paper elsewhere.
